src/device_grpc_impl.cpp

```diff
@@ -378,35 +366,32 @@ void AstarteDeviceGRPC::AstarteDeviceGRPCImpl::connection_attempt() {
   // least for that long.
   // See: https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_context.html
   std::unique_ptr<ClientContext> client_context = std::make_unique<ClientContext>();
-
-  // Call the attach RPC.
   std::unique_ptr<ClientReader<gRPCMessageHubEvent>> reader =
       stub_->Attach(client_context.get(), node);
 
-  // Start a thread for the event stream.
-  event_handler_ = std::thread(&AstarteDeviceGRPCImpl::handle_events, this,
-                               std::move(client_context), std::move(reader));
+  std::jthread event_handler(&AstarteDeviceGRPCImpl::handle_events, this, token,
+                             std::move(client_context), std::move(reader));
 
-  // Reset the stream error flag
   grpc_stream_error_.store(false);
 
   // Wait for the event stream to finish.
-  // N.B. This makes this function blocking!
-  if (event_handler_.joinable()) {
-    event_handler_.join();
+  if (event_handler.joinable()) {
+    event_handler.join();
   }
 }
 
 void AstarteDeviceGRPC::AstarteDeviceGRPCImpl::handle_events(
-    std::unique_ptr<grpc::ClientContext> context,
+    std::stop_token token, std::unique_ptr<grpc::ClientContext> context,
     std::unique_ptr<ClientReader<gRPCMessageHubEvent>> reader) {
   (void)context;
+  // the device is connected
+  connected_.store(true);
```

Comment:
At this point, we do not know if the device is connected. 
If the message hub does not exist, then the first call of the `reader->Read` function will return an error. If the connection has been achieved correctly, then the `reader->Read` function will block indefinitely.

Reply:
I would move this store operation to the parent thread. As (following the latest changes), the connection is completed there.